### PR TITLE
Feat/validation plotting

### PR DESCRIPTION
**Pull request by gtdang**
## Description
Scoring and plotting functions to measure model performance over cycles. 

See [examples/cycle/cycle_scoring.ipynb](https://github.com/AutoResearch/autora/blob/8d896f67f9c076161b257bcae7d8da05ff1bb778/example/cycle/cycle_scoring.ipynb) for general usage and examples. 
See [example/cycle/cycle_scoring_bms.ipynb](https://github.com/AutoResearch/autora/blob/0265d06f74e3c45053b0e7526400624c250550a4/example/cycle/cycle_scoring_bms.ipynb) for an example evaluating a BMS theorist cycle. 

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- Functions to return and plot scores of cycle models using a test/validation split. 
- Can use base estimator's default scorer or one of sklearn's [scoring metrics](https://scikit-learn.org/stable/modules/model_evaluation.html#the-scoring-parameter-defining-model-evaluation-rules).
- Jupyter notebook with example of logistic regression model validation using Iris toy dataset. 

### Remarks:
- No tests have been developed for these features yet. Will add them in January. 
- We'll probably need to test this with one of our theorists as well. I've only tested this with sklearn's `LogisticRegression`. 

### Screenshots:
Screen Shot 2022-12-23 at 10 31 28 AM

### CODE REVIEW

File: autora/cycle/__init__.py [hunk 1]
```diff
@@ -1,3 +1,9 @@
 from .simple import SimpleCycle as Cycle
 
-from .plot_utils import plot_results_panel_2d, plot_results_panel_3d  # isort:skip
+from .plot_utils import (  # isort:skip
```
Comment (hollandjg) [inline autora/cycle/__init__.py:3]: why do we have the isort:skip here? it shouldn't make a difference – the imports look to be imported in the right order.

Comment (gtdang) [reply autora/cycle/__init__.py:3]: Ah. It will try to sort with .plot_utils before .simple imports (because alphabetical). This causes an import error because the plot_utils imports Cycle. Perhaps plot_utils shouldn't import Cycle but instead import SimpleCycle?

Comment (hollandjg) [reply autora/cycle/__init__.py:3]: Ah, I see. A more fundamental question is whether we should import the members of .plot_utils at all, or whether you should just do this in your analysis script:
```python
from autora.cycle import Cycle  # for the default cycler, or
from autora.cycle.simple import SimpleCycle  # if you're wanting the SimpleCycle explicitly
from autora.cycle.plot import plot_cycle_score
```

File: autora/cycle/plot_utils.py [hunk 2]
```diff
@@ -472,3 +480,102 @@ def plot_results_panel_3d(
...
+
+    # Score each cycle
+    l_scores = []
+    for i, y_pred in d_y_pred.items():
+        l_scores.append(scorer(y_true, y_pred, **kwargs))
+
+    return l_scores
+
+
+def plot_cycle_score(
```
Comment (hollandjg) [inline autora/cycle/plot_utils.py:529]: I really like this interface! I went through the Notebook example, and it's really neat. You've managed to give the user the full power of all the scikit-learn metrics with very little effort. Nice.

File: autora/cycle/plot_utils.py [hunk 3]
```diff
@@ -472,3 +480,102 @@ def plot_results_panel_3d(
...
+    cycle: Cycle,
+    X: np.ndarray,
+    y_true: np.ndarray,
+    scorer: Optional[Callable] = None,
+    x_label: str = "Cycle",
+    y_label: Optional[str] = None,
+    figsize: Tuple[float, float] = (6.4, 4.8),
```
Comment (hollandjg) [inline autora/cycle/plot_utils.py:536]: Is there a way we can just inherit this from matplotlib? You could have some logic like: "if the user passes a figure themselves, then use that. If there's no figure passed in, then make one using the default parameters from matplotlib."

**Review by hollandjg — COMMENTED** [review]
This is awesome! It's a really cool, clean, small PR. 

I really like the interface, especially the fact that there's a bunch of complexity in the calculation hidden behind a really simple way of running the code. The notebook is a good example of using a finite dataset to simulate running experiments too.

I'd love to see the regression version of this! 

I've a couple of questions in comments above, but nothing show-stopping at all. Once you've got the tests in, this can totally be merged.

File: autora/cycle/__init__.py (hunk 1 above)
Comment (gtdang) [reply autora/cycle/__init__.py:3]: Yeah I suppose we could do it that way. I was trying to keep it similar to how we import plot utils for autora.theorist.darts but I'm not sure if there are examples of those plotters being used anywhere. The `__init__` for darts does have a `from .visualize import darts_model_plot`. I think that's for the model graph though. Thoughts?

File: autora/cycle/plot_utils.py (hunk 3 above)
Comment (gtdang) [reply autora/cycle/plot_utils.py]: Updated the script to access Matplotlib's rcParams to get the default figure size.

File: autora/cycle/plot_utils.py [hunk 4]
```diff
@@ -472,3 +480,102 @@ def plot_results_panel_3d(
...
+        if scorer is not None:
+            y_label = scorer.__name__
+        else:
+            y_label = "Score"
+    ax.set_ylabel(y_label)
+    ax.xaxis.set_major_locator(MaxNLocator(integer=True))
+
+    # Turn off spines
+    ax.spines.right.set_visible(False)
+    ax.spines.top.set_visible(False)
```
[the remark this answers is not in the document]
Comment (chadcwilliams) [reply autora/cycle/plot_utils.py:580]: I do agree with George that these spines should default off. Maybe it could be conformed to match how the plot_results_panel_2d function handles it (with spines as an argument set to false):

`if not spines:
    ax.spines.right.set_visible(False)
    ax.spines.top.set_visible(False)`

File: autora/cycle/plot_utils.py [hunk 5]
```diff
@@ -472,3 +481,102 @@ def plot_results_panel_3d(
...
+    scorer: Optional[Callable] = None,
+    x_label: str = "Cycle",
+    y_label: Optional[str] = None,
+    figsize: Tuple[float, float] = rcParams["figure.figsize"],
+    scorer_kw: dict = {},
+    plot_kw: dict = {},
+) -> plt.Figure:
+    """
+    Plots scoring metrics of cycle's theories given test data.
+    Args:
```
Comment (chadcwilliams) [inline autora/cycle/plot_utils.py:543]: I am wondering a bit about the y-limits. Here, it allows it to move freely based on the data, but this can be quite misleading when comparing different plots. For example, in example/cycle_scoring.ipynb you see some plots start at ~75% on cycle 1 and others at ~92% but because the y-limit is free to move, they look like similar progressions across cycles (despite the former gaining much more accuracy than the latter). 

I am wondering if you might find it useful to add some defaults (e.g., lower = 50%, higher = 100%) that can be modified by the user if necessary. Or even the lower bound can have thresholds so that it chooses 0%, 25%, or 50% depending where the lower bound of the data lies.

File: autora/cycle/plot_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gtdang) [reply autora/cycle/plot_utils.py:580]: Yeah that's the basic idea. At runtime, Matplotlib has all of its default plot parameters stored within an object called rcParams. This controls things like fonts, weights, and other things like which spines are showing. It initializes the object upon import. You can then edit the instance default configurations by changing the attributes stored in the rcParams object. For example: 

```python 
mpl.rcParams['axes.spines.top'] = False
mpl.rcParams['axes.spines.right'] = False
```

Would turn off all top and right spines for all future plots in your instance. 

So one solution would have those lines at the top  of the plot_utils file. That way whenever you import the plotters the rcParams will be set to our defaults. The user can always switch them back on if they know how to access rcParams.  

Alternatively if we wanted more control over the the styling we can create a custom style sheet for the project. Essentially we could create a custom style sheet containing all defaults specific to autora. You would just load the default styles with a command `plt.style.use('autora')` at the top of all of our plotting utilities. This could be bit overkill though. Custom styles sheets are often used when creating figures for a publication with specific style requirements. I've used them to standardize figure styles to conform to company branding.

File: autora/cycle/plot_utils.py (hunk 5 above)
Comment (gtdang) [reply autora/cycle/plot_utils.py:543]: Yeah that's a good idea. I think we can add a keyword for `ylim` and `xlim` so the user can specify if they want to zoom on a specific part of the plot.

Comment (gtdang) [reply autora/cycle/plot_utils.py:545]: Ok, I've added this feature. The last cell of the notebook example has been updated with an example of usage.

File: autora/cycle/plot_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gtdang) [reply autora/cycle/plot_utils.py]: I've refactored to remove the spine removal from within the plotting functions; And I added lines at the to top of the file to change the rcParam spines default to False. This occurs during module import and will turn off spines for all plots generated within the instance. 

I'm not sure if this is the best solution because it could lead to some unexpected behavior in some edge cases. Mostly it would be if a user was working in Jupyter notebook, the runtime instance stays open, they import the functions, and now any other plots they create will be spineless. They will just have to revert the rcParams or set a different style to achieve to bring back the spines. I don't think that's totally unreasonable?

[the remark this answers is not in the document]
Comment (chadcwilliams) [reply autora/cycle/plot_utils.py]: Yeah I think that's totally fine. I do not think it's too much to ask for them to learn a bit of matplotlib if they want to modify their plot design. And I definitely like putting the formatting elsewhere like John initially suggested, this feels cleaner and more streamlined. In time, we could think about adding more parameters there so that they are more easily accessible rather than having users write them out themselves. But I feel that is definitely out of the scope for the current PR.

File: autora/cycle/plot_utils.py (hunk 5 above)
Comment (chadcwilliams) [reply autora/cycle/plot_utils.py:543]: Awesome, looks good!

**Review by chadcwilliams — APPROVED** [review]
I am happy with this, it's looking great! At this point I have no more comments.

File: autora/cycle/__init__.py (hunk 1 above)
Comment (gtdang) [reply autora/cycle/__init__.py]: I changed the import statement in the plot_utils to import Cycle from simple directly. This avoids the import error and needing the isort:skip.
